File: app/src-tauri/src/hn.rs
use std::collections::HashSet;
use std::pin::Pin;
use std::sync::atomic::{AtomicUsize, Ordering};
use std::time::Duration;

use futures_util::future::join_all;
use serde::{Deserialize, Serialize};
use tokio::sync::Semaphore;

const HN_API_BASE: &str = "https://hacker-news.firebaseio.com/v0";
const USER_AGENT: &str =
    "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0 Safari/537.36";

/// Hard caps so a mega-thread (front-page Show HNs routinely clear 1000+
/// comments) can't turn one command into a thousand sequential HTTP round
/// trips or stall the UI — good enough to read the discussion without
/// waiting on every last nested reply.
const MAX_COMMENTS: usize = 300;
const MAX_DEPTH: u32 = 6;
const MAX_CONCURRENCY: usize = 12;

#[derive(Serialize, Clone)]
pub struct HnComment {
    pub id: i64,
    pub by: Option<String>,
    /// Sanitized HTML (HN comment bodies use a handful of tags: `<p>`, `<i>`, `<a>`, `<pre><code>`).
    pub text: String,
    pub time: Option<i64>,
    pub children: Vec<HnComment>,
}

#[derive(Deserialize)]
struct HnItemRaw {
    id: i64,
    by: Option<String>,
    text: Option<String>,
    time: Option<i64>,
    kids: Option<Vec<i64>>,
    deleted: Option<bool>,
    dead: Option<bool>,
    // Story-only fields (absent/None on comment items).
    title: Option<String>,
    url: Option<String>,
    score: Option<i64>,
    descendants: Option<i64>,
}

/// One entry in a ranked HN section list (New / Top / Best).
#[derive(Serialize, Clone)]
pub struct HnStorySummary {
    pub id: i64,
    pub title: String,
    /// The story's external link; for text-only posts (Ask HN etc. surfaced
    /// on these ranked lists) falls back to the HN discussion page itself.
    pub url: String,
    pub by: Option<String>,
    pub score: Option<i64>,
    pub time: Option<i64>,
    pub descendants: Option<i64>,
}

#[derive(Serialize)]
pub struct HnSectionPage {
    pub stories: Vec<HnStorySummary>,
    pub total: i64,
}

const HN_SECTIONS: &[(&str, &str)] = &[("new", "newstories"), ("top", "topstories"), ("best", "beststories")];

const ALGOLIA_SEARCH_URL: &str = "https://hn.algolia.com/api/v1/search";

#[derive(Serialize)]
pub struct HnSearchPage {
    pub stories: Vec<HnStorySummary>,
    pub page: i64,
    pub total_pages: i64,
}

#[derive(Deserialize)]
struct AlgoliaHit {
    #[serde(rename = "objectID")]
    object_id: String,
    title: Option<String>,
    url: Option<String>,
    author: Option<String>,
    points: Option<i64>,
    created_at_i: Option<i64>,
    num_comments: Option<i64>,
}

#[derive(Deserialize)]
struct AlgoliaResponse {
    hits: Vec<AlgoliaHit>,
    page: i64,
    #[serde(rename = "nbPages")]
    nb_pages: i64,
}

fn sanitize_comment(html: &str) -> String {
    let allowed_tags: HashSet<&str> = ["p", "i", "em", "b", "strong", "a", "pre", "code", "blockquote", "br"]
        .into_iter()
        .collect();
    ammonia::Builder::default()
        .tags(allowed_tags)
        .link_rel(Some("noopener noreferrer nofollow"))
        .clean(html)
        .to_string()
}

/// Deleted/flagged items come back as a bare JSON `null` rather than an
/// object, so the item itself (not just its fields) is optional.
async fn fetch_item(client: &reqwest::Client, id: i64) -> Result<Option<HnItemRaw>, String> {
    let url = format!("{HN_API_BASE}/item/{id}.json");
    let resp = client.get(&url).send().await.map_err(|e| e.to_string())?;
    let body = resp.text().await.map_err(|e| e.to_string())?;
    serde_json::from_str::<Option<HnItemRaw>>(&body).map_err(|e| e.to_string())
}

async fn fetch_id_list(client: &reqwest::Client, endpoint: &str) -> Result<Vec<i64>, String> {
    let url = format!("{HN_API_BASE}/{endpoint}.json");
    let resp = client.get(&url).send().await.map_err(|e| e.to_string())?;
    let body = resp.text().await.map_err(|e| e.to_string())?;
    serde_json::from_str::<Vec<i64>>(&body).map_err(|e| e.to_string())
}

async fn fetch_story(client: &reqwest::Client, sem: &Semaphore, id: i64) -> Option<HnStorySummary> {
    let permit = sem.acquire().await.ok()?;
    let item = fetch_item(client, id).await.ok().flatten()?;
    drop(permit);

    if item.deleted.unwrap_or(false) || item.dead.unwrap_or(false) {
        return None;
    }
    Some(HnStorySummary {
        id: item.id,
        title: item.title.unwrap_or_default(),
        url: item
            .url
            .unwrap_or_else(|| format!("https://news.ycombinator.com/item?id={}", item.id)),
        by: item.by,
        score: item.score,
        time: item.time,
        descendants: item.descendants,
    })
}

/// Fetch one page of a ranked Hacker News section — "new", "top", or "best",
/// mapping to HN's own /newest, /news, and /best. The id list is re-fetched
/// fresh on every call rather than cached/frozen across pages, so a live
/// ranking shift mid-scroll can shift or repeat an item — the same drift
/// HN's own /news?p=N pagination has, and not worth solving for a live,
/// unpersisted view.
#[tauri::command]
pub async fn fetch_hn_section(section: String, offset: i64, limit: i64) -> Result<HnSectionPage, String> {
    let endpoint = HN_SECTIONS
        .iter()
        .find(|(name, _)| *name == section)
        .map(|(_, endpoint)| *endpoint)
        .ok_or_else(|| format!("Unknown HN section: {section}"))?;

    let client = reqwest::Client::builder()
        .user_agent(USER_AGENT)
        .timeout(Duration::from_secs(15))
        .build()
        .map_err(|e| e.to_string())?;

    let ids = fetch_id_list(&client, endpoint).await?;
    let total = ids.len() as i64;
    let start = offset.clamp(0, total) as usize;
    let end = (offset.max(0) + limit.max(0)).clamp(0, total) as usize;

    let sem = Semaphore::new(MAX_CONCURRENCY);
    let futs: Vec<_> = ids[start..end]
        .iter()
        .map(|id| fetch_story(&client, &sem, *id))
        .collect();
    let stories = join_all(futs).await.into_iter().flatten().collect();

    Ok(HnSectionPage { stories, total })
}

/// Search Hacker News stories via Algolia's free HN Search API — the same
/// backend news.ycombinator.com/item's own "Search" box uses. Firebase (used
/// by `fetch_hn_section`) has no search of its own. Results are relevance-
/// ranked, restricted to stories (`tags=story`), and paginated by Algolia's
/// own page/nbPages rather than an offset.
#[tauri::command]
pub async fn search_hn(query: String, page: i64) -> Result<HnSearchPage, String> {
    if query.trim().is_empty() {
        return Ok(HnSearchPage { stories: Vec::new(), page: 0, total_pages: 0 });
    }

    let client = reqwest::Client::builder()
        .user_agent(USER_AGENT)
        .timeout(Duration::from_secs(15))
        .build()
        .map_err(|e| e.to_string())?;

    let resp = client
        .get(ALGOLIA_SEARCH_URL)
        .query(&[
            ("query", query.as_str()),
            ("tags", "story"),
            ("page", &page.max(0).to_string()),
        ])
        .send()
        .await
        .map_err(|e| format!("Request failed: {e}"))?;

    if !resp.status().is_success() {
        return Err(format!("Server returned {}", resp.status()));
    }

    let body = resp.text().await.map_err(|e| e.to_string())?;
    let parsed: AlgoliaResponse = serde_json::from_str(&body).map_err(|e| e.to_string())?;

    let stories = parsed
        .hits
        .into_iter()
        .filter_map(|hit| {
            let id: i64 = hit.object_id.parse().ok()?;
            Some(HnStorySummary {
                title: hit.title.unwrap_or_default(),
                url: hit
                    .url
                    .unwrap_or_else(|| format!("https://news.ycombinator.com/item?id={id}")),
                by: hit.author,
                score: hit.points,
                time: hit.created_at_i,
                descendants: hit.num_comments,
                id,
            })
        })
        .collect();

    Ok(HnSearchPage { stories, page: parsed.page, total_pages: parsed.nb_pages })
}

fn fetch_comment_tree<'a>(
    client: &'a reqwest::Client,
    sem: &'a Semaphore,
    id: i64,
    depth: u32,
    budget: &'a AtomicUsize,
) -> Pin<Box<dyn std::future::Future<Output = Option<HnComment>> + Send + 'a>> {
    Box::pin(async move {
        if budget.load(Ordering::Relaxed) == 0 {
            return None;
        }

        let permit = sem.acquire().await.ok()?;
        let item = fetch_item(client, id).await.ok().flatten()?;
        drop(permit);

        if item.deleted.unwrap_or(false) || item.dead.unwrap_or(false) {
            return None;
        }
        let text = item.text?; // no body (e.g. a poll option) — nothing to show
        if budget
            .fetch_update(Ordering::Relaxed, Ordering::Relaxed, |b| b.checked_sub(1))
            .is_err()
        {
            return None;
        }

        let children = if depth < MAX_DEPTH {
            match item.kids {
                Some(kids) => {
                    let futs: Vec<_> = kids
                        .iter()
                        .map(|kid| fetch_comment_tree(client, sem, *kid, depth + 1, budget))
                        .collect();
                    join_all(futs).await.into_iter().flatten().collect()
                }
                None => Vec::new(),
            }
        } else {
            Vec::new()
        };

        Some(HnComment {
            id: item.id,
            by: item.by,
            text: sanitize_comment(&text),
            time: item.time,
            children,
        })
    })
}

/// Fetch a Hacker News discussion's comments (recursively, via HN's free
/// Firebase item API) given the story's item id — captured from an RSS
/// entry's `hn_item_id` (see `rss.rs::extract_hn_item_id`) or read straight
/// off a `news.ycombinator.com/item?id=` URL. Capped at MAX_COMMENTS /
/// MAX_DEPTH so a mega-thread can't stall the UI.
#[tauri::command]
pub async fn fetch_hn_comments(story_id: i64) -> Result<Vec<HnComment>, String> {
    let client = reqwest::Client::builder()
        .user_agent(USER_AGENT)
        .timeout(Duration::from_secs(15))
        .build()
        .map_err(|e| e.to_string())?;

    let story = fetch_item(&client, story_id)
        .await?
        .ok_or_else(|| "Story not found".to_string())?;
    let kids = story.kids.unwrap_or_default();
    let budget = AtomicUsize::new(MAX_COMMENTS);
    let sem = Semaphore::new(MAX_CONCURRENCY);

    let futs: Vec<_> = kids
        .iter()
        .map(|kid| fetch_comment_tree(&client, &sem, *kid, 0, &budget))
        .collect();
    Ok(join_all(futs).await.into_iter().flatten().collect())
}

File: mobile_tanwords/src/services/hn.ts
/**
 * Hacker News API — mobile port of app/core/src/hn.rs. Same endpoints:
 * ranked sections via the official Firebase API, story search via Algolia,
 * threaded comments via the official API with a shared budget + depth cap.
 * Comment text is sanitized to the same allowlist ammonia used in Rust.
 *
 * Exported names are camelCase (searchHn / fetchHnSection / fetchHnComments);
 * field names match the desktop serde output exactly: HnSearchPage keeps
 * `total_pages` (no rename in hn.rs) and comment/story fields keep snake_case
 * none → all plain words, so both match.
 */

const HN_API_BASE = "https://hacker-news.firebaseio.com/v0";
const ALGOLIA_SEARCH_URL = "https://hn.algolia.com/api/v1/search";
const USER_AGENT =
  "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0 Safari/537.36";

/** Same caps as the desktop: mega-threads can't stall the UI. */
const MAX_COMMENTS = 300;
const MAX_DEPTH = 6;
const MAX_CONCURRENCY = 12;
const REQUEST_TIMEOUT_MS = 15_000;

export interface HnComment {
  id: number;
  by: string | null;
  /** Sanitized HTML (allowlist: p,i,em,b,strong,a,pre,code,blockquote,br). */
  text: string;
  time: number | null;
  children: HnComment[];
}

export interface HnStorySummary {
  id: number;
  title: string;
  /** Story link; Ask-HN-style text posts fall back to the discussion page. */
  url: string;
  by: string | null;
  score: number | null;
  time: number | null;
  descendants: number | null;
}

export interface HnSectionPage {
  stories: HnStorySummary[];
  total: number;
}

export interface HnSearchPage {
  stories: HnStorySummary[];
  page: number;
  total_pages: number; // eslint-disable-line camelcase -- matches desktop serde output
}

export type HnSection = "new" | "top" | "best";

const HN_SECTIONS: Record<HnSection, string> = {
  new: "newstories",
  top: "topstories",
  best: "beststories",
};

interface HnItemRaw {
  id: number;
  by?: string | null;
  text?: string | null;
  time?: number | null;
  kids?: number[] | null;
  deleted?: boolean | null;
  dead?: boolean | null;
  title?: string | null;
  url?: string | null;
  score?: number | null;
  descendants?: number | null;
}

/* ---------- small shared helpers ---------- */

async function fetchJson(url: string, init?: RequestInit): Promise<unknown> {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), REQUEST_TIMEOUT_MS);
  try {
    const resp = await fetch(url, {
      ...init,
      headers: { "User-Agent": USER_AGENT, ...(init?.headers ?? {}) },
      signal: controller.signal,
    });
    if (!resp.ok) {
      throw new Error(`Server returned HTTP ${resp.status} for ${url}`);
    }
    const text = await resp.text();
    if (!text) return null;
    return JSON.parse(text) as unknown;
  } catch (e) {
    if (e instanceof Error && e.name === "AbortError") {
      throw new Error(`Request to ${url} timed out after ${REQUEST_TIMEOUT_MS / 1000}s`);
    }
    throw e;
  } finally {
    clearTimeout(timer);
  }
}

/** Deleted/flagged items come back as bare JSON null. */
async function fetchItem(id: number): Promise<HnItemRaw | null> {
  const parsed = await fetchJson(`${HN_API_BASE}/item/${id}.json`);
  return parsed as HnItemRaw | null;
}

async function fetchIdList(endpoint: string): Promise<number[]> {
  const parsed = await fetchJson(`${HN_API_BASE}/${endpoint}.json`);
  return Array.isArray(parsed) ? (parsed as number[]) : [];
}

/* ---------- comment HTML sanitization (ammonia allowlist parity) ---------- */

const ALLOWED_TAGS = new Set(["p", "i", "em", "b", "strong", "a", "pre", "code", "blockquote", "br"]);

/** Whitelist-based sanitizer mirroring ammonia's builder on the desktop.
 *  Keeps allowed tags (with `rel="noopener noreferrer nofollow"` on links),
 *  unwraps anything else but keeps its text, escapes entities. */
function sanitizeComment(html: string): string {
  const decoded = decodeEntities(html);
  let out = "";
  let i = 0;
  while (i < decoded.length) {
    const lt = decoded.indexOf("<", i);
    if (lt === -1) {
      out += escapeHtml(decoded.slice(i));
      break;
    }
    out += escapeHtml(decoded.slice(i, lt));
    const gt = decoded.indexOf(">", lt);
    if (gt === -1) {
      out += escapeHtml(decoded.slice(lt));
      break;
    }
    const tag = decoded.slice(lt + 1, gt);
    const m = /^\s*(\/?)\s*([a-zA-Z0-9]+)/.exec(tag);
    if (!m) {
      i = gt + 1;
      continue;
    }
    const closing = m[1] === "/";
    const name = m[2].toLowerCase();
    if (ALLOWED_TAGS.has(name)) {
      if (name === "br") {
        out += "<br>";
      } else if (name === "a") {
        if (closing) {
          out += "</a>";
        } else {
          const hrefMatch = /\bhref\s*=\s*("([^"]*)"|'([^']*)'|([^\s>]+))/i.exec(tag);
          const href = hrefMatch?.[2] ?? hrefMatch?.[3] ?? hrefMatch?.[4] ?? "";
          const safe = /^https?:\/\//i.test(href);
          out += safe
            ? `<a href="${escapeAttr(href)}" rel="noopener noreferrer nofollow">`
            : "<a>";
        }
      } else {
        out += closing ? `</${name}>` : `<${name}>`;
      }
    }
    // Disallowed tags are unwrapped: their inner text is still emitted,
    // matching ammonia's default behavior (keep content, drop the element).
    i = gt + 1;
  }
  return out;
}

function decodeEntities(input: string): string {
  return input
    .replace(/&#(\d+);/g, (_, n) => {
      const cp = Number(n);
      try {
        return Number.isFinite(cp) ? String.fromCodePoint(cp) : "";
      } catch {
        return "";
      }
    })
    .replace(/&#x([0-9a-fA-F]+);/g, (_, n) => {
      const cp = parseInt(n, 16);
      try {
        return Number.isFinite(cp) ? String.fromCodePoint(cp) : "";
      } catch {
        return "";
      }
    })
    .replace(/&(nbsp|amp|lt|gt|quot|apos);/g, (_, name: string) =>
      name === "nbsp"
        ? " "
        : name === "amp"
          ? "&"
          : name === "lt"
            ? "<"
            : name === "gt"
              ? ">"
              : name === "quot"
                ? '"'
                : "'"
    );
}

function escapeHtml(input: string): string {
  return input.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;");
}

function escapeAttr(input: string): string {
  return escapeHtml(input).replace(/"/g, "&quot;");
}

/* ---------- concurrency pool ---------- */

async function mapLimit<T, R>(
  items: readonly T[],
  limit: number,
  worker: (item: T) => Promise<R>
): Promise<R[]> {
  const results = new Array<R>(items.length);
  let cursor = 0;
  async function run(): Promise<void> {
    while (cursor < items.length) {
      const idx = cursor++;
      results[idx] = await worker(items[idx]);
    }
  }
  const workers = Array.from({ length: Math.min(limit, items.length) }, () => run());
  await Promise.all(workers);
  return results;
}

/* ---------- public API ---------- */

/**
 * One page of a ranked HN section ("new" | "top" | "best"), mirroring
 * `fetch_hn_section`. The id list is re-fetched per call, so live ranking
 * drift is the same trade-off as the desktop.
 */
export async function fetchHnSection(
  section: HnSection,
  offset: number,
  limit: number
): Promise<HnSectionPage> {
  const endpoint = HN_SECTIONS[section];
  if (!endpoint) throw new Error(`Unknown HN section: ${section}`);
  const ids = await fetchIdList(endpoint);
  const total = ids.length;
  const start = Math.max(0, Math.min(offset, total));
  const end = Math.max(start, Math.min(start + Math.max(0, limit), total));
  const slice = ids.slice(start, end);

  const stories = await mapLimit(slice, MAX_CONCURRENCY, async (id): Promise<HnStorySummary | null> => {
    const item = await fetchItem(id).catch(() => null);
    if (!item || item.deleted || item.dead) return null;
    return {
      id: item.id,
      title: item.title ?? "",
      url: item.url ?? `https://news.ycombinator.com/item?id=${item.id}`,
      by: item.by ?? null,
      score: item.score ?? null,
      time: item.time ?? null,
      descendants: item.descendants ?? null,
    };
  });

  return { stories: stories.filter((s): s is HnStorySummary => s !== null), total };
}

/** Algolia story search, mirroring `search_hn`. */
export async function searchHn(query: string, page: number): Promise<HnSearchPage> {
  if (!query.trim()) {
    return { stories: [], page: 0, total_pages: 0 };
  }
  const params = new URLSearchParams({
    query,
    tags: "story",
    page: String(Math.max(0, page)),
  });
  const parsed = (await fetchJson(`${ALGOLIA_SEARCH_URL}?${params.toString()}`)) as {
    hits?: Array<{
      objectID?: string;
      title?: string | null;
      url?: string | null;
      author?: string | null;
      points?: number | null;
      created_at_i?: number | null;
      num_comments?: number | null;
    }> | null;
    page?: number;
    nbPages?: number;
  } | null;

  const stories: HnStorySummary[] = (parsed?.hits ?? [])
    .map((hit): HnStorySummary | null => {
      const id = Number(hit.objectID);
      if (!Number.isInteger(id) || id <= 0) return null;
      return {
        id,
        title: hit.title ?? "",
        url: hit.url ?? `https://news.ycombinator.com/item?id=${id}`,
        by: hit.author ?? null,
        score: hit.points ?? null,
        time: hit.created_at_i ?? null,
        descendants: hit.num_comments ?? null,
      };
    })
    .filter((s): s is HnStorySummary => s !== null);

  return {
    stories,
    page: parsed?.page ?? Math.max(0, page),
    total_pages: parsed?.nbPages ?? 0,
  };
}

/** Threaded comments for a story, mirroring `fetch_hn_comments`. */
export async function fetchHnComments(storyId: number): Promise<HnComment[]> {
  const story = await fetchItem(storyId);
  if (!story) throw new Error("Story not found");
  const kids = story.kids ?? [];
  let budget = MAX_COMMENTS;

  async function fetchTree(id: number, depth: number): Promise<HnComment | null> {
    if (budget <= 0) return null;
    const item = await fetchItem(id).catch(() => null);
    if (!item || item.deleted || item.dead) return null;
    const text = item.text;
    if (!text) return null; // no body (e.g. a poll option)
    budget -= 1;

    let children: HnComment[] = [];
    if (depth < MAX_DEPTH && item.kids && item.kids.length > 0) {
      const childResults = await mapLimit(item.kids, MAX_CONCURRENCY, (kid) =>
        fetchTree(kid, depth + 1)
      );
      children = childResults.filter((c): c is HnComment => c !== null);
    }

    return {
      id: item.id,
      by: item.by ?? null,
      text: sanitizeComment(text),
      time: item.time ?? null,
      children,
    };
  }

  const topLevel = await mapLimit(kids, MAX_CONCURRENCY, (kid) => fetchTree(kid, 0));
  return topLevel.filter((c): c is HnComment => c !== null);
}
